src/server-logger.ts

```diff
@@ -0,0 +1,257 @@
+  private logBuffer: Array<ILogBufferItem> = [];
+  private pendingHttpRequest: boolean = false;
+  private debounceLogUploadTime: number;
+  private debounceTimer: any = null;
+
+
+  constructor (config: ILoggerConfig, logger: Logger) {
+    this.config = config;
+    this.logger = logger;
+    /* if we have all the needed config options, set this up */
+    if (
+      !config.url ||
+      !config.accessToken ||
+      !config.appVersion
+    ) {
+      const errMessage = 'Lacking necessary config options to set up server logging. ' +
+        'Not sending logs to server for this logger instance';
+      this.logger.error(errMessage, { ...config }, true);
+      throw new Error(errMessage);
+    }
+
+    this.isInitialized = true;
+    this.debounceLogUploadTime = config.uploadDebounceTime || DEFAULT_UPLOAD_DEBOUNCE;
+
+    window.addEventListener('unload', this.sendAllLogsInstantly.bind(this));
+  }
+
+  public addLogToSend (logLevel: LogLevel, message: string | Error, details: IDetail | IDetail[]): void {
+    if (!this.isInitialized) {
+      return;
+    }
+
+    if (message instanceof Error) {
+      details = details || message;
+      message = message.message;
+    }
+
+    let logMessage = this.convertToLogMessage(message, details);
+    let trace = this.convertToTrace(logLevel, logMessage);
+    let traceMessageSize = calculateLogMessageSize(trace);
+
+    /* if the individual message exceeds the max allowed size, truncate it */
+    if (traceMessageSize > MAX_LOG_SIZE) {
+      const newTrace = this.truncateLog(logLevel, logMessage);
+      this.logger.warn('log message too large to send to server. truncating log details and/or message', {
+        originalTrace: (trace as any),
+        truncatedTrace: (newTrace as any)
+      }, true);
+
+      /* newTrace will be `null` if the truncated trace was still too big */
+      if (newTrace === null) {
+        this.logger.error('truncated message is still too large to send to server. not sending message', {
+          originalTrace: (trace as any)
+        }, true);
+        return;
+      }
+
+      /* set the trace to our new trunctated trace item */
+      traceMessageSize = calculateLogMessageSize(newTrace);
+      trace = newTrace;
+    }
+
+    /* use the last item in the buffer if it exists, otherwise start with a blank buffer item */
+    let bufferItem: ILogBufferItem;
+    if (this.logBuffer.length) {
+      bufferItem = this.logBuffer[this.logBuffer.length - 1];
+    } else {
+      bufferItem = {
+        size: 0,
+        traces: []
+      };
+    }
+
+    /* if pushing our trace onto the buffer item will be too large, we need a new buffer item */
+    const exceedsMaxLogSize = bufferItem.size + traceMessageSize > MAX_LOG_SIZE;
+    if (exceedsMaxLogSize) {
+      // console.log('%c `exceedsMaxLogSize` was `true`', 'color: #32a852', { logBuffer: [...this.logBuffer], bufferItem, incomingTrace: trace, incomingTraceSize: traceMessageSize }); // TODO: delete me
```

Comment:
If you uncomment all these `console.log('%c ...')` statements and then run `npm start`, you can test out the logger.